lib/assist/constants.go

```diff
@@ -0,0 +1,35 @@
+/*
+ Copyright 2023 Gravitational, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package assist
+
+// MessageClasses contains type of assist message we expect users to send.
+// When running on Cloud we attempt to classify user messages in one of those
+// categories. If this succeeds, we send an event into the analytics pipeline.
+//
+// Keys are the category names, those are the ones reported in the event and
+// generated by the model. Values are the category description. They are used to
+// build the model prompt and allow to provide more context to the model to
+// improve the classification.
+var MessageClasses = map[string]string{
+	"command execution": "the user want to execute a command on one or many servers",
+	"troubleshooting":   "the user wants to diagnose a problem or understand an error message",
+	"configuration":     "the user wants to generate configuration for a software which is not Teleport",
```

Comment:
How can the LLM know if the software is "Teleport"? Also, do we even care about Teleport/non-Teleport scenario?

Reply:
The prompt contains a super-short Teleport description. From what I observed this is enough for gpt-4 to reliably infer if the request is about Teleport or not. (gpt-3.5 showed some issues but was right most of the time). I suspect gpt-4 also learned what Teleport is and can leverage this information when classifying.

I think this is important to separate Teleport-related requests from non-teleport related requests as those two categories are not actionable in the same way. We can help the model to answer Teleport configuration requests by giving it access to the docs and allowing it to link back the user to the docs. On the other hand, we rely on the generic part of the model to answer most other configuration questions (embedding man pages might help but this is a long shot).

From what I understand we will want to know if users are asking "write me a working nginx configuration" or if they are asking "I want the teleport agent configuration to provide access to a ssh server and a database at the same time".